Mathlib/RingTheory/HopfAlgebra.lean

```diff
@@ -115,3 +133,146 @@ instance toHopfAlgebra : HopfAlgebra R R where
+instance instHopfAlgebraStruct : HopfAlgebraStruct R A[G] where
+  antipode := Finsupp.lsum R fun g => Finsupp.lsingle (-g) ∘ₗ antipode
+
+@[simp]
+lemma antipode_single (g : G) (a : A) :
+    antipode (R := R) (single g a) = single (-g) (antipode (R := R) a) := by
+  simp only [AddMonoidAlgebra, antipode, Finsupp.coe_lsum, map_zero, Finsupp.sum_single_index]
+  rfl
+
+instance instHopfAlgebra : HopfAlgebra R A[G] where
+  toHopfAlgebraStruct := instHopfAlgebraStruct R A G
+  mul_antipode_rTensor_comul := lhom_ext fun a b => by
+    simp only [LinearMap.coe_comp, Function.comp_apply, comul_single,
+      ← (Coalgebra.Repr.arbitrary R b).eq, map_sum, TensorProduct.map_tmul, lsingle_apply,
+      LinearMap.rTensor_tmul, antipode_single, LinearMap.mul'_apply, single_mul_single,
+      neg_add_cancel, counit_single, Algebra.linearMap_apply, coe_algebraMap]
+    simp only [← lsingle_apply (k := R), ← map_sum (lsingle R A (0 : G)),
+      HopfAlgebra.sum_antipode_mul_eq_smul, Algebra.algebraMap_eq_smul_one]
+  mul_antipode_lTensor_comul := lhom_ext fun a b => by
+    simp only [LinearMap.coe_comp, Function.comp_apply, comul_single,
+      ← (Coalgebra.Repr.arbitrary R b).eq, map_sum, TensorProduct.map_tmul,
+      lsingle_apply, LinearMap.lTensor_tmul, antipode_single, LinearMap.mul'_apply,
+      single_mul_single, add_neg_cancel, counit_single, Algebra.linearMap_apply, coe_algebraMap]
+    simp only [← lsingle_apply (k := R), ← map_sum (lsingle R A (0 : G)),
+      HopfAlgebra.sum_mul_antipode_eq_smul, Algebra.algebraMap_eq_smul_one]
+
+end AddMonoidAlgebra
+
+namespace LaurentPolynomial
+
+open Finsupp
+
+variable (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] [HopfAlgebra R A]
+
+instance instHopfAlgebra : HopfAlgebra R A[T;T⁻¹] :=
+  inferInstanceAs (HopfAlgebra R <| AddMonoidAlgebra A ℤ)
+
+variable {R A}
+
+@[simp]
+theorem comul_C (a : A) :
+    Coalgebra.comul (R := R) (C a)
+      = TensorProduct.map (lsingle 0) (lsingle 0) (Coalgebra.comul (R := R) a) :=
+  comul_single _ _ _ _ _
+
+@[simp]
+theorem comul_T (n : ℤ) :
+    Coalgebra.comul (R := R) (T (R := A) n) = T n ⊗ₜ[R] T n := by
+  erw [comul_single]
+  simp [Algebra.TensorProduct.one_def]
+
+@[simp]
+theorem comul_C_mul_T (a : A) (n : ℤ) :
+    Coalgebra.comul (R := R) (C a * T n)
+      = TensorProduct.map (lsingle n) (lsingle n) (Coalgebra.comul (R := R) a) := by
+  rw [← single_eq_C_mul_T, comul_single]
+
+@[simp high]
+theorem comul_C_mul_T_self (a : R) (n : ℤ) :
+    Coalgebra.comul (R := R) (C a * T n) = T n ⊗ₜ[R] (C a * T n) := by
+  erw [comul_C_mul_T]
+  simp
+
+@[simp]
+theorem counit_C (a : A) :
+    Coalgebra.counit (R := R) (C a) = Coalgebra.counit (R := R) a :=
+  counit_single _ _ _ _ _
+
+@[simp]
+theorem counit_T (n : ℤ) :
+    Coalgebra.counit (R := R) (T (R := A) n) = 1 := by
+  erw [counit_single]
+  rw [Bialgebra.counit_one]
+
+theorem counit_C_mul_T (a : A) (n : ℤ) :
+    Coalgebra.counit (R := R) (C a * T n) = Coalgebra.counit (R := R) a := by
+  simp
```

Comment:
Don't these theorems belong in the file about Coalgebras? I think they only need `Coalgebra R A`).

Reply:
Thanks, done - I've moved the API for the coalgebra structure on monoid algebras to the coalgebra file too.